File: 1857-largest-color-value-in-a-directed-graph/1857-largest-color-value-in-a-directed-graph.py
class Solution:
    def largestPathValue(self, colors: str, edges: List[List[int]]) -> int:
        n = len(colors)
        graph = defaultdict(list)
        indegree = [0] * n

        # 建圖 + 入度
        for u, v in edges:
            graph[u].append(v)
            indegree[v] += 1

        # dp[i][c]：到達節點 i，顏色 c（0~25）最多出現了幾次
        dp = [[0] * 26 for _ in range(n)]

        queue = deque()
        for i in range(n):
            if indegree[i] == 0:
                queue.append(i)
        
        # 初始：每個節點的自己顏色先+1
        for i in range(n):
            color_index = ord(colors[i]) - ord('a')
            dp[i][color_index] = 1

        visited = 0
        max_color_value = 0

        while queue:
            node = queue.popleft()
            visited += 1
            for neighbor in graph[node]:
                for c in range(26):
                    dp[neighbor][c] = max(dp[neighbor][c], dp[node][c] + (1 if c == ord(colors[neighbor]) - ord('a') else 0))
                indegree[neighbor] -= 1
                if indegree[neighbor] == 0:
                    queue.append(neighbor)
            max_color_value = max(max_color_value, max(dp[node]))

        # 有 cycle
        return -1 if visited < n else max_color_value

File: 1857-largest-color-value-in-a-directed-graph/1857-largest-color-value-in-a-directed-graph.cpp
class Solution {
public:
    int largestPathValue(string colors, vector<vector<int>>& edges) {
        int n = colors.size();
        vector<vector<int>> graph(n);
        vector<int> indegree(n, 0);

        // 建立圖與計算入度
        for (auto& edge : edges) {
            int u = edge[0], v = edge[1];
            graph[u].push_back(v);
            indegree[v]++;
        }

        // 建立 dp 陣列，dp[i][c] 表示從起點到 i，顏色 c 出現的最大次數
        vector<vector<int>> dp(n, vector<int>(26, 0));
        
        queue<int> q;
        // 將所有入度為 0 的節點放入 queue
        for (int i = 0; i < n; ++i) {
            if (indegree[i] == 0)
                q.push(i);
            dp[i][colors[i] - 'a'] = 1; // 初始化：自己的顏色出現一次
        }

        int visited = 0;
        int maxColorCount = 0;

        while (!q.empty()) {
            int node = q.front(); q.pop();
            visited++;
            maxColorCount = max(maxColorCount, *max_element(dp[node].begin(), dp[node].end()));

            for (int neighbor : graph[node]) {
                for (int c = 0; c < 26; ++c) {
                    int add = (colors[neighbor] - 'a' == c) ? 1 : 0;
                    dp[neighbor][c] = max(dp[neighbor][c], dp[node][c] + add);
                }
                indegree[neighbor]--;
                if (indegree[neighbor] == 0)
                    q.push(neighbor);
            }
        }

        // 如果還有節點沒被處理，代表圖中有環
        return visited == n ? maxColorCount : -1;
    }
};